### ISSUE TITLE
querystring.parse('name&value=12')

### ISSUE THREAD

**Issue by drsm** [nginx/njs#349]
```js
> var qs = require('querystring'), o = qs.parse('name&value=12'); console.log(o)
[Object: null prototype] { name: '', value: '12' }
> new URLSearchParams('name&value=12')
URLSearchParams { 'name' => '', 'value' => '12' }
```
vs.
```js
>> var qs = require('querystring'), o = qs.parse('name&value=12'); console.log(o)
{name&value:'',value:'12'}
undefined
```

**Comment by drsm** [comment]
Also, according to the [spec](https://url.spec.whatwg.org/#application/x-www-form-urlencoded) `querystring.parse()`  should do the following:
> Replace any 0x2B (+) in name and value with 0x20 (SP).

```js
> querystring.parse('space+separated+key=space+seaprated+value')
[Object: null prototype] {
  'space separated key': 'space seaprated value'
}
```

**Comment by xeioex** [comment]
@drsm 

>Also, according to the spec, querystring.parse() should do the following:
>>Replace any 0x2B (+) in name and value with 0x20 (SP).

We already [do](https://github.com/nginx/njs/blob/master/src/njs_query_string.c#L105) it.

**Comment by drsm** [comment]
@xeioex 

> @drsm
> 
> > Also, according to the spec, querystring.parse() should do the following:
> > > Replace any 0x2B (+) in name and value with 0x20 (SP).
> 
> We already [do](https://github.com/nginx/njs/blob/master/src/njs_query_string.c#L105) it.

Oh, just didn't notice it in the source. Actually, it is a small compatibility problem. The spec is a real mess here.

```js
> var id = (x) => x;
undefined
> require('querystring').parse('space+separated+key=space+seaprated+value', '&', '=', { decodeURIComponent: id })
[Object: null prototype] {
  'space%20separated%20key': 'space%20seaprated%20value'
}
```
vs.
```js
>> var id = (x) => x;
undefined
>> require('querystring').parse('space+separated+key=space+seaprated+value', '&', '=', { decodeURIComponent: id })
{
 space+separated+key: 'space+seaprated+value'
}
```

**Comment by xeioex** [comment]
@drsm 

Take a look

```diff
# HG changeset patch
# User Dmitry Volyntsev <xeioex@nginx.com>
# Date 1604563843 0
#      Thu Nov 05 08:10:43 2020 +0000
# Node ID d3faf0a79e5c7841f5dbff0b7a01143ba8ec3682
# Parent  c791e3943df1974a300e5f112d0108004b1835c6
Fixed querystring.parse().

The issue happened when the first eq symbol is located after the
separator, whereas it should be looked for only in the string segment
before the separator.

This fixes #349 issue on Github.

diff --git a/src/njs_query_string.c b/src/njs_query_string.c
--- a/src/njs_query_string.c
+++ b/src/njs_query_string.c
@@ -269,7 +269,7 @@ njs_query_string_match(u_char *p, u_char
         return p;
     }
 
-    while (p < (end - length)) {
+    while (p <= (end - length)) {
         if (memcmp(p, v->start, length) == 0) {
             return p;
         }
@@ -402,7 +402,7 @@ njs_query_string_parse(njs_vm_t *vm, njs
             goto next;
         }
 
-        val = njs_query_string_match(key, end, &eq);
+        val = njs_query_string_match(key, part, &eq);
 
         size = val - key;
 
diff --git a/src/test/njs_unit_test.c b/src/test/njs_unit_test.c
--- a/src/test/njs_unit_test.c
+++ b/src/test/njs_unit_test.c
@@ -17987,6 +17987,11 @@ static njs_unit_test_t  njs_test[] =
       njs_str("{freespace:''}") },
 
     { njs_str("var qs = require('querystring');"
+              "var obj = qs.parse('name&value=12');"
+              "njs.dump(obj)"),
+      njs_str("{name:'',value:'12'}") },
+
+    { njs_str("var qs = require('querystring');"
               "var obj = qs.parse('baz=fuz&muz=tax', 'fuz');"
               "njs.dump(obj)"),
       njs_str("{baz:'',&muz:'tax'}") },
```
